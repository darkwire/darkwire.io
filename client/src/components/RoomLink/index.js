import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Copy } from 'react-feather'
import Clipboard from 'clipboard'
import $ from 'jquery';

class RoomLink extends Component {
  constructor(props) {
    super(props)
    this.state = {
      roomUrl: `${window.location.origin}/${props.roomId}`,
    }
  }

  componentDidMount() {
    const clip = new Clipboard('.copy-room')

    clip.on('success', () => {
      $('.copy-room').tooltip('show')
      setTimeout(() => {
        $('.copy-room').tooltip('hide')
      }, 3000)
    })

    $(() => {
      $('.copy-room').tooltip({
        trigger: 'manual',
      })
    })
  }

  componentWillUnmount() {
    $('.copy-room').tooltip('hide')
  }

  render() {
    return (
      <form>
        <div className="form-group">
          <div className="input-group">
            <input id="room-url" className="form-control" type="text" readOnly value={this.state.roomUrl} />
            <div className="input-group-append">
              <button
                className="copy-room btn btn-secondary"
                type="button"
                data-toggle="tooltip"
                data-placement="bottom"
                data-clipboard-text={this.state.roomUrl}
                title="Copied!"
              >
                <Copy />
              </button>
            </div>
          </div>
        </div>
      </form>
    )
  }
}

RoomLink.propTypes = {
  roomId: PropTypes.string.isRequired,
}

export default RoomLink
