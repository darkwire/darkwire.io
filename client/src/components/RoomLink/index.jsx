import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Copy } from 'react-feather';
import Clipboard from 'clipboard';
import $ from 'jquery';

const RoomLink = ({ roomId, translations }) => {
  const roomUrl = `${window.location.origin}/${roomId}`;

  React.useEffect(() => {
    const clip = new Clipboard('.copy-room');

    clip.on('success', () => {
      $('.copy-room').tooltip('show');
      setTimeout(() => {
        $('.copy-room').tooltip('hide');
      }, 3000);
    });

    $(() => {
      $('.copy-room').tooltip({
        trigger: 'manual',
      });
    });

    return () => {
      if ($('.copy-room').tooltip) $('.copy-room').tooltip('hide');
    };
  }, []);

  return (
    <form>
      <div className="form-group">
        <div className="input-group">
          <input id="room-url" className="form-control" type="text" readOnly value={roomUrl} />
          <div className="input-group-append">
            <button
              className="copy-room btn btn-secondary"
              type="button"
              data-toggle="tooltip"
              data-placement="bottom"
              data-clipboard-text={roomUrl}
              title={translations.copyButtonTooltip}
            >
              <Copy />
            </button>
          </div>
        </div>
      </div>
    </form>
  );
};

RoomLink.propTypes = {
  roomId: PropTypes.string.isRequired,
  translations: PropTypes.object.isRequired,
};

export default RoomLink;
