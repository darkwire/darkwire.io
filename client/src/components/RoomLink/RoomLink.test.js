import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import RoomLink from '.';
import mock$ from 'jquery';

const mockTooltip = jest.fn().mockImplementation(param => {
  // console.log('tooltip', param);
});

jest.mock('jquery', () => {
  return jest.fn().mockImplementation(param => {
    // console.log('$', param);
    if (typeof param === 'function') {
      param();
    }
    return {
      tooltip: mockTooltip,
    };
  });
});

jest.useFakeTimers();

const mockTranslations = {
  copyButtonTooltip: 'copyButton',
};

describe('RoomLink', () => {
  afterEach(() => {
    mock$.mockClear();
  });

  it('should display', async () => {
    const { asFragment, unmount } = render(<RoomLink roomId="roomId" translations={mockTranslations} />);

    expect(asFragment()).toMatchSnapshot();

    expect(mock$).toHaveBeenLastCalledWith('.copy-room');
    expect(mockTooltip).toHaveBeenLastCalledWith({ trigger: 'manual' });
    mock$.mockClear();

    unmount();

    expect(mock$).toHaveBeenLastCalledWith('.copy-room');
    expect(mockTooltip).toHaveBeenLastCalledWith('hide');
  });

  it('should copy link', async () => {
    // Mock execCommand for paste
    document.execCommand = jest.fn(() => true);

    const { getByTitle } = render(<RoomLink roomId="roomId" translations={mockTranslations} />);

    fireEvent.click(getByTitle(mockTranslations.copyButtonTooltip));

    expect(document.execCommand).toHaveBeenLastCalledWith('copy');
    expect(mock$).toHaveBeenCalledTimes(4);
    expect(mock$).toHaveBeenLastCalledWith('.copy-room');
    expect(mockTooltip).toHaveBeenLastCalledWith('show');

    // Wait for tooltip to close
    jest.runAllTimers();

    expect(mock$).toHaveBeenCalledTimes(6);
    expect(mock$).toHaveBeenLastCalledWith('.copy-room');
    expect(mockTooltip).toHaveBeenLastCalledWith('hide');
  });
});
