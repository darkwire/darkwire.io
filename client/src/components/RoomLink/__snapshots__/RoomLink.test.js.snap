// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomLink should display 1`] = `
<DocumentFragment>
  <form>
    <div
      class="form-group"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control"
          id="room-url"
          readonly=""
          type="text"
          value="http://localhost/roomId"
        />
        <div
          class="input-group-append"
        >
          <button
            class="copy-room btn btn-secondary"
            data-clipboard-text="http://localhost/roomId"
            data-placement="bottom"
            data-toggle="tooltip"
            title="copyButton"
            type="button"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="13"
                rx="2"
                ry="2"
                width="13"
                x="9"
                y="9"
              />
              <path
                d="M5 15H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h9a2 2 0 0 1 2 2v1"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
